metricbeat/module/apache/status/status.go

```diff
@@ -48,46 +61,97 @@ func (m *MetricSeter) Setup(ms *helper.MetricSet) error {
 		return err
 	}
 
-	m.ServerStatusPath = config.ServerStatusPath
-	m.Username = config.Username
-	m.Password = config.Password
-	if m.Password != "" && m.Username != "" {
-		m.Authentication = true
-	} else {
-		m.Authentication = false
+	m.URLs = make(map[string]string, len(ms.Config.Hosts))
+
+	// Parse the config, create URLs, and check for errors.
+	var errs multierror.Errors
+	for _, host := range ms.Config.Hosts {
+		u, err := getURL(config.Username, config.Password, config.ServerStatusPath, host)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		m.URLs[host] = u.String()
+		debugf("apache-status URL=%s", redactPassword(*u))
 	}
 
-	return nil
+	return errs.Err()
 }
 
+// Fetch makes an HTTP request to fetch status metrics from the mod_status
+// endpoint.
 func (m *MetricSeter) Fetch(ms *helper.MetricSet, host string) (event common.MapStr, err error) {
+	url, ok := m.URLs[host]
+	if !ok {
+		return nil, fmt.Errorf("url not found for host '%s'", host)
+	}
 
-	u, err := url.Parse(host + m.ServerStatusPath)
+	req, err := http.NewRequest("GET", url, nil)
+	client := &http.Client{}
+	resp, err := client.Do(req)
 	if err != nil {
-		logp.Err("Invalid Apache HTTPD server-status page: %v", err)
-		return nil, err
+		return nil, fmt.Errorf("error making HTTP request: %v", err)
 	}
+	defer resp.Body.Close()
```

Comment:
According to the docs for [`client.Do()`](https://golang.org/pkg/net/http/#Client.Do), when err is nil, `resp` always contains a non-nil `resp.Body`. So `defer resp.Body.Close()` should come after the err check to prevent a nil pointer deference no `resp`.